client/src/main/scala/org/http4s/client/middleware/FollowRedirect.scala

```diff
@@ -2,54 +2,147 @@ package org.http4s
 package client
 package middleware
 
+import org.http4s.Method._
+import org.http4s.headers._
+import org.http4s.util.string._
 import scalaz.Kleisli
 import scalaz.concurrent.Task
+import scalaz.stream.Process._
+import scalaz.syntax.monad._
+import scodec.bits.ByteVector
 
-/** Follow redirect responses */
+/** 
+  * Client middleware to follow redirect responses.
+  * 
+  * A 301 or 302 response is followed by:
+  * - a GET if the request was GET or POST
+  * - a HEAD if the request was a HEAD
+  * - the original request method and body if the body had no effects
+  * - the redirect is not followed otherwise
+  * 
+  * A 303 response is followed by:
+  * - a HEAD if the request was a HEAD
+  * - a GET for all other methods
+  * 
+  * A 307 or 308 response is followed by:
+  * - the original request method and body, if the body had no effects
+  * - the redirect is not followed otherwise
+  * 
+  * Whenever we follow with a GET or HEAD, an empty body is sent, and
+  * all payload headers defined in https://tools.ietf.org/html/rfc7231#section-3.3
+  * are stripped.
+  * 
+  * If the response does not contain a valid Location header, the redirect is
+  * not followed.
+  */
 object FollowRedirect {
 
+  // TODO move this somewhere more sensible in 0.15, but we don't want to break
+  // bincompat in a bugfix for 0.14.
+  //
+  // https://tools.ietf.org/html/rfc7231#section-3.3
+  private val PayloadHeaderKeys = Set(
```

Comment:
I'm surprised this wasn't a bigger list.  What does a `Content-Type` mean, for instance?